pkg/metallb/frr.go

```diff
@@ -0,0 +1,548 @@
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof(
+		"Allows the FRR nodes to receive all routes in namespace %s with mode type: all",
+		builder.Definition.Namespace)
+
+	if builder.errorMsg != "" {
+		return builder
+	}
+
+	// Check if the BGP routers exist
+	if len(builder.Definition.Spec.BGP.Routers) == 0 {
+		builder.errorMsg = "No BGP routers defined."
+
+		return builder
+	}
+
+	// Iterate through routers to find the one with the desired ASN (if applicable) and update its neighbors
+	for i, router := range builder.Definition.Spec.BGP.Routers {
+		for j := range router.Neighbors {
+			// Update the ToReceive mode to AllowAll for each neighbor
+			builder.Definition.Spec.BGP.Routers[i].Neighbors[j].ToReceive.Allowed.Mode = frrtypes.AllowAll
+		}
+	}
+
+	return builder
+}
+
+// WithToReceiveModeFiltered allows only specified prefixes to be received.
+func (builder *FrrConfigurationBuilder) WithToReceiveModeFiltered(prefixes []string,
+	neighborIndex int) *FrrConfigurationBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof(
+		"Allows the FRR nodes to receive specific routes in namespace %s in mode type: filtered",
+		builder.Definition.Namespace)
+
+	if builder.errorMsg != "" {
+		return builder
+	}
+
+	// Ensure the neighbor index is within bounds
+	if neighborIndex < 0 || neighborIndex >= len(builder.Definition.Spec.BGP.Routers[0].Neighbors) {
+		builder.errorMsg = fmt.Sprintf("Neighbor index %d is out of bounds", neighborIndex)
+
+		return builder
+	}
+
+	// Validate CIDR prefixes
+	for _, prefix := range prefixes {
+		if _, _, err := net.ParseCIDR(prefix); err != nil {
+			glog.V(100).Infof("the frrConfiguration prefix %s is not a valid CIDR", prefix)
+			builder.errorMsg = fmt.Sprintf("The prefix %s is not a valid CIDR", prefix)
+
+			return builder
+		}
+	}
+
+	// Prepare the list of allowed prefixes
+	allowedPrefixes := make([]frrtypes.PrefixSelector, len(prefixes))
+
+	for i, prefix := range prefixes {
+		glog.V(100).Infof("AllowMode 'filtered' receives only routes within the prefix %v", prefix)
+		allowedPrefixes[i] = frrtypes.PrefixSelector{Prefix: prefix}
+	}
+
+	// Update the specific neighbor's ToReceive field
+	builder.Definition.Spec.BGP.Routers[0].Neighbors[neighborIndex].ToReceive.Allowed.Prefixes = allowedPrefixes
+
+	return builder
+}
+
+// WithBGPNeighbor defines a single neighbor IP and ASN number.
+func (builder *FrrConfigurationBuilder) WithBGPNeighbor(bgpPeerIP string, remoteAS uint32) *FrrConfigurationBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof(
+		"Define BGP Neighbor %s with peer IP %s and ASN %v",
+		builder.Definition.Name, bgpPeerIP, remoteAS)
+
+	if bgpPeerIP == "" {
+		glog.V(100).Infof("The BGP peer IP should not be blank")
+
+		builder.errorMsg = fmt.Sprintf("the BGP peer IP %s is an empty string", bgpPeerIP)
+	}
+
+	if net.ParseIP(bgpPeerIP) == nil {
```

Comment:
I bellieve this validation also covers previous one -  bgpPeerIP is empty

Reply:
Yes